src/binary_annotator.cpp

```diff
@@ -181,23 +192,78 @@ void BinaryAnnotator::BuildVTable(uint64_t offset,
       return;
     }
 
-    const uint16_t value = GetScalar<uint16_t>(field_offset);
+    const uint16_t offset_from_table = GetScalar<uint16_t>(field_offset);
+
+    VTable::Entry entry;
+    entry.field = field;
+    entry.offset_from_table = offset_from_table;
+    fields.insert(std::make_pair(field->id(), entry));
+
+    std::string default_label;
+    if (offset_from_table == 0) {
+      // Not present, so could be default or be optional.
+      if (field->required()) {
+        // If this is a required field, make it known this is an error.
+        regions.push_back(MakeBinaryRegion(
+            field_offset, sizeof(uint16_t), BinaryRegionType::VOffset, 0, 0,
+            std::string("ERROR: required field `") + field->name()->c_str() +
+                "` (id: " + std::to_string(field->id()) + ") is not present!"));
+        return;
+      } else {
+        // Its an optional field, so get the default value and interpret and
+        // provided an annotation for it.
+        if (IsScalar(field->type()->base_type())) {
+          default_label += " <defaults to ";
+          default_label += IsFloat(field->type()->base_type())
+                               ? std::to_string(field->default_real())
+                               : std::to_string(field->default_integer());
+          default_label += "> (";
+        } else {
+          default_label += " <null> (";
+        }
+        default_label +=
+            reflection::EnumNameBaseType(field->type()->base_type());
+        default_label += ")";
+      }
+    }
 
-    fields.push_back(std::make_pair(field, GetScalar<uint16_t>(field_offset)));
     regions.push_back(MakeBinaryRegion(
         field_offset, sizeof(uint16_t), BinaryRegionType::VOffset, 0, 0,
         std::string("offset to field `") + field->name()->c_str() +
-            "` (id: " + std::to_string(field->id()) + ")" +
-            ((value == 0) ? " <defaults to " +
-                                std::to_string(field->default_integer()) + ">"
-                          : "")));
+            "` (id: " + std::to_string(field->id()) + ")" + default_label));
+
+    fields_processed++;
   });
 
-  sections_.insert(std::make_pair(
-      vtable_offset,
-      MakeBinarySection(table->name()->str(), BinarySectionType::VTable,
-                        std::move(regions))));
-  vtables_.insert(std::make_pair(vtable_offset, VTable{ std::move(fields) }));
+  // Check if we covered all the expcentant fields. If not, we need to add them
+  // as unknown fields.
+  const uint16_t expectant_vtable_fields =
+      (vtable_size - sizeof(uint16_t) - sizeof(uint16_t)) / sizeof(uint16_t);
+
+  for (uint16_t id = fields_processed; id < expectant_vtable_fields; ++id) {
```

Comment:
optional nit: this might be a little clearer if there was just one for loop that branched if the schema did not recognize the field id